---
title: "R Notebook"
---


```{r, include=FALSE}
library(tidyverse)
library(factoextra)
```

```{r, include=FALSE}
path_to_deseq_model_results_dir = '~/Desktop/tmp/deseq_model_20200911/deseq_model/results'
model_directory_list = list.dirs(path_to_deseq_model_results_dir, full.names = TRUE, recursive = FALSE)

full_model_and_ruvr_results_list = model_directory_list[grepl('*libraryprotocol_librarydate_genotype*', model_directory_list)]

protein_coding_gene_and_markers = read_csv('/home/chase/Desktop/tmp/deseq_model_20200911/deseq_model/data/KN99_protein_coding_gene_list_with_markers.csv', col_names = 'gene_id')

log2_norm_counts = read_csv('/home/chase/Desktop/tmp/deseq_model_20200911/deseq_model/results/libraryprotocol_librarydate_genotype_model_with_results/log2_norm_counts.csv')
log2_norm_counts = cbind(log2_norm_counts, protein_coding_gene_and_markers)

meta_qual_df = read_csv('/home/chase/Desktop/tmp/deseq_model_20200911/deseq_model/data/filtered_metadata_20200812.csv')

log2_norm_residual_list
log2_norm_residuals = cbind(log2_norm_residuals, protein_coding_gene_and_markers)

# de results
results_names = as.character(lapply(full_model_and_ruvr_results_list, basename))
de_results_tables = lapply(full_model_and_ruvr_results_list, function(x) file.path(x, 'de_results'))
de_results_tables = lapply(de_results_tables, list.files, '*csv', full.names=TRUE)
names(de_results_tables) = results_names

# list of residual files
residuals_list = list.files(full_model_and_ruvr_results_list, 'log2_norm_space_residuals*', full.names = TRUE, recursive = TRUE)
```

```{r}

# create scree plot for each residuals
# writeScreePlotToFile = function(residual_csv_path){
#   residual_df = read_csv(residual_csv_path)
#   
#   pc_object = prcomp(residual_df)
#   
#   dir_path = dirname(residual_csv_path)
#   plot_title = basename(dir_path)
#   scree_output_path = file.path(dir_path, 'screeplot.pdf')
#   
#   pdf(scree_output_path)
#     plot(fviz_eig(pc_object, main=plot_title, ylim = c(0,100)))
#   dev.off()
# }
# 
# lapply(residuals_list, writeScreePlotToFile)
```

```{r}
createListResultsDataframes = function(de_results_table_list, protein_coding_gene_and_markers){
 # read in list of .csv as list of dataframes
 de_results_table_df_list =  lapply(de_results_tables, read_csv)
 # add gene id col
 de_results_table_list = lapply(de_results_table_df_list, cbind, protein_coding_gene_and_markers)
 
 return(de_results_table_df_list)
}

geneListFilteredMetaExpression = function(expression_df, meta_qual_df, gene_filter_vector){
   #' meant to filter either norm counts or residuals (any expression matrix) for genes in gene_filter_vector (eg c('CKF44_02132', CKF44_12313', ...))

  # get list of columns in expression_df without the gene_id column
  samples = colnames(expression_df)[colnames(expression_df) != 'gene_id']
  
  filtered_expression_df = expression_df %>% filter(gene_id %in% gene_filter_vector)
  
  meta_expression_df = filtered_expression_df %>% 
    gather(FASTQFILENAME, expression, samples) %>%
    left_join(meta_qual_df)

  return(meta_expression_df)
  
}

plotLibraryProtocolDateGenotype = function(meta_expression_df){
  #' expects the quant column to be named 'expression'

  g1 = ggplot(meta_expression_df) +
  geom_jitter(aes(x = gene_id, y = expression, color=LIBRARYPROTOCOL))
  plot(g1)
  
  g2 = ggplot(meta_expression_df) +
  geom_jitter(aes(x = gene_id, y = expression, color=LIBRARYPROTOCOL))
  plot(g2)
  
  g3 = ggplot(resid_meta_expression_df) +
  geom_jitter(aes(x = gene_id, y = expression, color=GENOTYPE), shape=GENOTYPE == 'CNAG_0000O') + theme(legend.position = "none")
  plot(g3)
  
  }
```

```{r, include=FALSE}

# create dataframe of counts of genes with adjp < .05 and > .95
pval_summary_df = tibble(model=NaN, genotype=NaN, num_less_05=NaN, num_greater_95=NaN, inbetween=NaN)
for(model in names(de_results_tables)){
  csv_list = de_results_tables[[model]]
  model = str_match(model, '(?<=genotype_).*')[1]
  for(csv_path in csv_list){
    genotype = str_match(basename(csv_path), 'GENOTYPECNAG_\\d+')[1]
    
    df = read_csv(csv_path)
    df = df[!is.na(df$padj), ] # remove rows that were removed due to low count (or outlier status?)
    num_less_05    = nrow(df[df$padj < .05,])
    num_greater_95 = nrow(df[df$padj > .95,])
    inbetween = nrow(df[(df$padj > .05 & df$padj < .95),])
    
    append_df = tibble(model=model, genotype=genotype, num_less_05 = num_less_05, 
                       num_greater_95 = num_greater_95, inbetween = inbetween)
    
    pval_summary_df = rbind(pval_summary_df, append_df)
  }
}
pval_summary_df = pval_summary_df[-1,]
level_order = c('model_with_results', 'k_1', 'k_2', 'k_3', 'k_4', 'k_5', 'k_6', 'k_7', 'k_8', 'k_9', 'k_10')
pval_summary_df$model = factor(pval_summary_df$model, levels= level_order)
pval_summary_df$total = pval_summary_df$num_less_05 + pval_summary_df$num_greater_95 + pval_summary_df$inbetween
```

```{r}
# plot num of p values in each genotype
# plot num of p values in each genotype
g = ggplot(pval_summary_df)+
  geom_boxplot(aes(model, num_less_05))+
  ylim(0,7000)
plot(g)

g1 = ggplot(pval_summary_df)+
  geom_boxplot(aes(model, num_less_05))+
  ylim(0,1000)
plot(g1)

g3 = ggplot(pval_summary_df)+
  geom_boxplot(aes(model, num_greater_95))+
  ylim(0,7000)
plot(g3)

g4 = ggplot(pval_summary_df)+
  geom_boxplot(aes(model, inbetween))+
  ylim(0,7000)
plot(g4)

g5 = ggplot(pval_summary_df)+
  geom_boxplot(aes(model, total))+
  ylim(0,7000)
plot(g5)
```


filter all tables for pvalue < .05, lfc > 2
``` {r}
# filter for lfc > 2, padj < .05
lfc_pval_filtered_results_table_list = lapply(de_results_table_list, filter, log2FoldChange > 2, padj < .05) 
# get 90th percentile of the above genes, across all tables, mean expression
mean_expression_list = lapply(lfc_pval_filtered_results_table_list, select, baseMean)
mean_expression_vector = bind_rows(mean_expression_list)
ninety_percentile = quantile(mean_expression_vector$baseMean, .9)

# filter results_table_list, but this time include filter on baseMean
filtered_results_table_list = lapply(de_results_table_list, filter, log2FoldChange > 2, padj < .05, baseMean > ninety_percentile)

filtered_gene_list = lapply(filtered_results_table_list, select, gene_id)
concat_filtered_gene_list = bind_rows(filtered_gene_list)
summary_filtered_gene_list = concat_filtered_gene_list %>% group_by(gene_id) %>% summarize(count=n())

# filter results_table_listt, no effect
no_effect_filtered_results_table_list = lapply(de_results_table_list, filter, abs(log2FoldChange) < .05, baseMean > ninety_percentile)

no_effect_filtered_gene_list = lapply(no_effect_filtered_results_table_list, select, gene_id)
no_effect_concat_filtered_gene_list = bind_rows(no_effect_filtered_gene_list)
no_effect_summary_filtered_gene_list = no_effect_concat_filtered_gene_list %>% group_by(gene_id) %>% summarize(count=n())
```
The two genes found to be variable in the most samples were: CKF44_03722, a RAN binding protein (RAN is a gene identified as producing a protein involved in transport into/out of the nucleus) and CKF44_05662, involved in sugar transport

genes that show little effect in most samples:
CKF44_00006 -- Prohibitin (highly conserved, ubiquitously expressed)
CFK44_07668 -- HAL protein kinase

```{r}
# filter log2_norm_counts for these four genes
# genes_of_interest = c('CKF44_03722', 'CKF44_05662','CKF44_00006','CKF44_07668')
# 
# filtered_log2_norm_counts = log2_norm_counts %>% filter(gene_id %in% genes_of_interest)
# meta_expression_df = filtered_log2_norm_counts %>% 
#   gather(FASTQFILENAME, log2_norm_count, colnames(log2_norm_counts)[1:640]) %>%
#   left_join(meta_qual_df)
# 
# filtered_log2_norm_residuals = log2_norm_residuals %>% filter(gene_id %in% genes_of_interest)
# resid_meta_expression_df = filtered_log2_norm_residuals %>%
#   gather(FASTQFILENAME, log2_norm_count, colnames(log2_norm_counts)[1:640])%>%
#   left_join(meta_qual_df)


```

```{r}
# plot

meta_expression_df$gene_id = as.factor(meta_expression_df$gene_id)
meta_expression_df$log2_norm_count = as.double(meta_expression_df$log2_norm_count)

g = ggplot(meta_expression_df) +
  geom_jitter(aes(x = gene_id, y = log2_norm_count, color=LIBRARYPROTOCOL))

g = ggplot(resid_meta_expression_df) +
  geom_jitter(aes(x = gene_id, y = log2_norm_count, color=as.factor(GENOTYPE))) + theme(legend.position = "none")


geno_filetered_meta_expression_df = meta_expression_df %>% filter(GENOTYPE == 'CNAG_00000')

g = plot_ly(x=geno_filetered_meta_expression_df$gene_id, y=geno_filetered_meta_expression_df$LIBRARYPROTOCOL, z=geno_filetered_meta_expression_df$log2_norm_count, type='scatter3d', color=geno_filetered_meta_expression_df$GENOTYPE)
plot(g )

```

```{r}

log2_norm_residuals = read_csv('/mnt/htcf_scratch/chasem/misc_scripts/deseq_model/results/libraryprotocol_librarydate_genotype_k_1/log2_norm_space_residuals.csv')
log2_norm_residuals = cbind(log2_norm_residuals, protein_coding_gene_and_markers)

log2_norm_residuals_no_gene_id = log2_norm_residuals %>% select(-gene_id)
resid_pc = prcomp(log2_norm_residuals_no_gene_id)
g1 = plot(fviz_eig(resid_pc, main = 'log2 norm residual pcs'))
resid_svd = svd(log2_norm_residuals_no_gene_id)
g2 = barplot(resid_svd$d[1:640], names.arg = 1:640)
plot(g1)
plot(g2)
```
```{r}

log2_norm_residuals = read_csv('/mnt/htcf_scratch/chasem/misc_scripts/deseq_model/results/libraryprotocol_librarydate_genotype_k_7/log2_norm_space_residuals.csv')
log2_norm_residuals = cbind(log2_norm_residuals, protein_coding_gene_and_markers)

log2_norm_residuals_no_gene_id = log2_norm_residuals %>% select(-gene_id)
resid_pc = prcomp(log2_norm_residuals_no_gene_id)
g1 = plot(fviz_eig(resid_pc, main = 'log2 norm residual pcs'))
resid_svd = svd(log2_norm_residuals_no_gene_id)
g2 = barplot(resid_svd$d[1:640], names.arg = 1:640)
plot(g1)
plot(g2)

```